router/dataplane.go

```diff
@@ -1390,12 +1390,22 @@ func (p *scionPacketProcessor) reset() (err error) {
 // @ ensures  respr.OutPkt !== rawPkt && respr.OutPkt != nil ==>
 // @ 	sl.AbsSlice_Bytes(respr.OutPkt, 0, len(respr.OutPkt))
 // @ ensures  reserr != nil ==> reserr.ErrorMem()
+// contracts for IO-spec
+// @ requires dp.Valid()
+// @ requires acc(ioLock.LockP(), _) && ioLock.LockInv() == SharedInv!< dp, ioSharedArg !>;
+// @ requires acc(&p.ingressID)
+// @ requires let absPkt := absIO_val(dp, rawPkt, p.ingressID) in
+// @	absPkt.isIO_val_Pkt2 ==> ElemWitness(ioSharedArg.IBufY, ifsToIO_ifs(p.ingressID), absPkt.IO_val_Pkt2_2)
+// @ ensures acc(&p.ingressID)
+// @ ensures reserr == nil && newAbsPkt.isIO_val_Pkt2 ==> ElemWitness(ioSharedArg.OBufY, newAbsPkt.IO_val_Pkt2_1, newAbsPkt.IO_val_Pkt2_2)
+// @ ensures reserr == nil && newAbsPkt.isIO_val_Pkt2 ==> newAbsPkt == absIO_val(dp, respr.OutPkt, respr.EgressID)
```

Comment:
we should break these lines to make them easier to read

Reply:
Additionally, I wonder if `reserr == nil && newAbsPkt.isIO_val_Pkt2 ==> newAbsPkt == absIO_val(dp, respr.OutPkt, respr.EgressID)` should be `reserr == nil ==> newAbsPkt.isIO_val_Pkt2 && newAbsPkt == absIO_val(dp, respr.OutPkt, respr.EgressID)` instead. Shouldn't we always return an `IO_val_Pkt2` when `reserr` is `nil`?

Reply:
Not necessarily; The condition `newAbsPkt.isIO_val_Pkt2` is true only when we receive a SCION packet. For instance, if we receive an EPIC, `newAbsPkt.isIO_val_Pkt2` will not be true, even if `reserr` is `nil`.